server/keyspace_service.go

```diff
@@ -0,0 +1,141 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"path"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/tikv/pd/server/keyspace"
+	"github.com/tikv/pd/server/storage/endpoint"
+	"go.etcd.io/etcd/clientv3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// KeyspaceServer wraps GrpcServer to provide keyspace service.
+type KeyspaceServer struct {
+	*GrpcServer
+}
+
+// getErrorHeader returns corresponding ResponseHeader based on err.
+func (s *KeyspaceServer) getErrorHeader(err error) *pdpb.ResponseHeader {
+	switch err {
+	case keyspace.ErrKeyspaceExists:
+		return s.wrapErrorToHeader(pdpb.ErrorType_DUPLICATED_ENTRY, err.Error())
+	case keyspace.ErrKeyspaceNotFound:
+		return s.wrapErrorToHeader(pdpb.ErrorType_ENTRY_NOT_FOUND, err.Error())
+	default:
+		return s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN, err.Error())
+	}
+}
+
+// validateRequest checks that server is serving and cluster id matches the requested.
+func (s *KeyspaceServer) validateRequest(header *pdpb.RequestHeader) error {
+	if s.IsClosed() {
+		return ErrNotStarted
+	}
+	if header.GetClusterId() != s.clusterID {
+		return status.Errorf(codes.FailedPrecondition, "mismatch cluster id, need %d but got %d", s.clusterID, header.GetClusterId())
+	}
+	return nil
+}
+
+// LoadKeyspace load and return target keyspace metadata.
+// Request must specify keyspace name.
+// On Error, keyspaceMeta in response will be nil,
+// error information will be encoded in response header with corresponding error type.
+func (s *KeyspaceServer) LoadKeyspace(_ context.Context, request *keyspacepb.LoadKeyspaceRequest) (*keyspacepb.LoadKeyspaceResponse, error) {
+	if err := s.validateRequest(request.GetHeader()); err != nil {
+		return nil, err
+	}
+	rc := s.GetRaftCluster()
+	if rc == nil {
+		return &keyspacepb.LoadKeyspaceResponse{Header: s.notBootstrappedHeader()}, nil
+	}
+
+	manager := s.GetKeyspaceManager()
+	meta, err := manager.LoadKeyspace(request.Name)
+	if err != nil {
+		return &keyspacepb.LoadKeyspaceResponse{Header: s.getErrorHeader(err)}, err
+	}
+	return &keyspacepb.LoadKeyspaceResponse{
+		Header:   s.header(),
+		Keyspace: meta,
+	}, nil
+}
+
+// WatchKeyspaces captures and sends keyspace metadata changes to the client via gRPC stream.
+// Note: It sends all existing keyspaces as it's first package to the client.
+func (s *KeyspaceServer) WatchKeyspaces(request *keyspacepb.WatchKeyspacesRequest, stream keyspacepb.Keyspace_WatchKeyspacesServer) error {
+	if err := s.validateRequest(request.GetHeader()); err != nil {
+		return err
+	}
+	rc := s.GetRaftCluster()
+	if rc == nil {
+		return stream.Send(&keyspacepb.WatchKeyspacesResponse{Header: s.notBootstrappedHeader()})
```

Comment:
When will it return a response or just an error?

Reply:
I think a rule of thumb is that if the error has a pdpb.ErrorType (e.g., cluster not bootstrapped), it should be send back in the stream.
Otherwise it should return error.
A example of this can be seen in [`ReportBuckets`](https://github.com/tikv/pd/blob/0791fe818b13398c723b7d6214afe230ed429f6a/server/grpc_service.go#L791)